### ISSUE TITLE
validation fails on master when run in parallel

### ISSUE THREAD

**Issue by davidssmith** [davidssmith/DCEMRI.jl#33]
Not sure if this is related to the new fitting methods or some breaking of the parallel environment across versions.

**Comment by notZaki** [comment]
I'm able to run the validation in parallel for the master on Julia 0.5.2 by doing:  
```
addprocs(3)
using DCEMRI
validate()
```
I get a decent speed improvement by running in parallel.
For example, noisy QIBA 6 validation went from ~26 seconds (one worker) to ~14 seconds (three workers).

In practice, I would replace `addprocs(3)` with:
```
import Hwloc
topology = Hwloc.topology_load()
counts = Hwloc.histmap(topology)
nCores = counts[:Core]
if nprocs()<nCores
  addprocs(nCores-nprocs()) # numWorkers = numCores - 1, my preference
  # addprocs(nCores-nprocs()+1) # Use all the cores.
end
```
so that the number of workers is determined by the user's CPU.

**Comment by davidssmith** [comment]
Mine runs, but the parameters for Ktrans=0.2 and the two highest vp (lower right corner) for the noiseless QIBA v4 phantom are wrong. This only happens when I run in parallel. 

I like your solution for replacing `addprocs`.  If you want to make a PR, I'll merge it. If you do, don't forget to add `Hwloc` to `REQUIRE`.

On the other hand, your PR with the new fitting method made things so fast that the overhead of dealing with workers is starting to be significant. I might look at just trying to multithread when possible, and then remove the worker stuff. Thoughts?

**Comment by notZaki** [comment]
I tried it, and that problem occurs when there are 4 workers, but not when there are 3, 2, or 1 workers - I didn't encounter the problem earlier because I either use 1 or 3 workers.

The noiseless QIBA 4 data has 90 pixels which are divided equally among all the workers. 
When using four workers, 90/4 = 22 with remainder = 2, and those remainder pixels probably don't get fitted at all.
The log says:
```
fitting 661 x 22 points on each of 4 workers
```
which means only 88 pixels were being fitted on. 
This would explain why both kt and vp are zero in those two pixels.
I wonder if a previous implementation accounted for the remainder pixels or if this bug was always there.

As for replacing `addprocs` with the `Hwloc` code, I'm not sure how that would work. 
We have to start the workers before loading the package, so... the package itself can't include the code that starts the workers, can it? (Previously, we had `require()`, but that's gone now) 

I'm not sure how much overhead the workers are adding but I think the current master works well. By default it uses a single worker. If the user wants multithreading, then the only thing they have to do is start the workers with `addprocs` before loading the package. 
I've included the run time for noisy QIBA 4 below to show that there is value in using multithreading - particularly when going from 1 to 2 workers, followed by diminishing returns. 

**Aside: Runtimes for noisy QIBA 4 using different number of workers**  
Processing 661 x 9000 points  

**numWorkers**|**runtime [s]**
-----|-----
1|41.9
2|27.5
3|24.6
4|22.7

- Runtimes from only one run, so take with grain of salt

**Comment by davidssmith** [comment]
Good catch. I think that bug has been around since I switched to dividing up the work into contiguous blocks, rather than interleaving worker assignments with something like `workeridxs =
 [workerid:nworkers:npoints]`. The latest commit should fix this issue.

You're point about `Hwloc` is well taken. I miss `require` *sniff*.  

I'm thinking that the new speed you've achieved makes plain threading more attractive. Not sure how to do that. Maybe the parallelization can be offloaded to BLAS routines by reformulating the fitting as a big matrix problem.

Also a GPU solution like `ArrayFire.jl` could work, but it makes it harder for the casual user to get up and running.

**Comment by notZaki** [comment]
There is the linearized version of the tofts models which can be described as a matrix problem and fitted with linear least squares.  
For example, the noisy QIBA 4 test took ~42 to ~23 seconds depending on the number of workers, while the linearized model takes less than 1.2 seconds on a single worker. 
The speedup comes at the cost of accuracy - mostly at the lowest ve region.

I've made a 'linearModels' branch on my cloned repository which implements the linearized tofts models.
The 'test' folder includes the text-log from the validation test along with the script used to do the test. 
I implemented this a long time ago, but never really figured out how to merge them into the `models=[1,2,3]` paradigm.
